File: .agents/skills/agent-skill-authoring/scripts/validate_codex_skill.py
from __future__ import annotations

import argparse
from pathlib import Path

import yaml


def _resolve_skill_path(value: str) -> Path:
    candidate = Path(value).expanduser()
    if candidate.exists():
        return candidate.resolve()
    repo_skill = Path(__file__).resolve().parents[4] / ".agents" / "skills" / value
    return repo_skill.resolve()


def _parse_frontmatter(content: str) -> dict[str, str]:
    if not content.startswith("---\n"):
        raise ValueError("SKILL.md must start with YAML frontmatter")
    _, frontmatter, _ = content.split("---", 2)
    parsed = yaml.safe_load(frontmatter.strip())
    if not isinstance(parsed, dict):
        raise ValueError("frontmatter must parse to a mapping")
    return parsed


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Validate a repository-local Codex skill."
    )
    parser.add_argument(
        "skill_path",
        help="Path to a skill directory or a skill name under the repo root .agents/skills/.",
    )
    args = parser.parse_args()

    skill_path = _resolve_skill_path(args.skill_path)
    skill_file = skill_path / "SKILL.md"
    if not skill_file.exists():
        raise SystemExit(f"missing SKILL.md at {skill_file}")

    frontmatter = _parse_frontmatter(skill_file.read_text(encoding="utf-8"))
    name = frontmatter.get("name")
    description = frontmatter.get("description")
    if not name:
        raise SystemExit("frontmatter is missing name")
    if not description:
        raise SystemExit("frontmatter is missing description")
    if name != skill_path.name:
        raise SystemExit(
            f"skill name '{name}' does not match directory '{skill_path.name}'"
        )

    print(f"skill: {name}")
    print(f"description: {description}")
    if (skill_path / "references").is_dir():
        print("references: present")
    if (skill_path / "scripts").is_dir():
        print("scripts: present")


if __name__ == "__main__":
    main()

File: .agents/skills/agent-skill-authoring/scripts/validate_codex_skill.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

type SkillFrontmatter = {
  description: string;
  name: string;
};

function main(): void {
  const skillArgument = process.argv[2];
  if (skillArgument === undefined || skillArgument.trim() === "") {
    usage("Missing skill path or name.");
  }

  const skillPath = resolveSkillPath(skillArgument);
  const skillFile = resolve(skillPath, "SKILL.md");
  if (!existsSync(skillFile)) {
    throw new Error(`missing SKILL.md at ${skillFile}`);
  }

  const frontmatter = parseFrontmatter(readFileSync(skillFile, "utf8"));
  const directoryName = skillPath.split(/[\\/]/u).at(-1);
  if (frontmatter.name !== directoryName) {
    throw new Error(
      `skill name '${frontmatter.name}' does not match directory '${directoryName ?? ""}'`,
    );
  }

  console.log(`skill: ${frontmatter.name}`);
  console.log(`description: ${frontmatter.description}`);
  if (isDirectory(resolve(skillPath, "references"))) {
    console.log("references: present");
  }
  if (isDirectory(resolve(skillPath, "scripts"))) {
    console.log("scripts: present");
  }
}

function resolveSkillPath(value: string): string {
  const directPath = resolve(value);
  if (existsSync(directPath)) {
    return directPath;
  }
  return resolve(defaultSkillsRoot(), value);
}

function parseFrontmatter(content: string): SkillFrontmatter {
  const match = /^---\r?\n([\s\S]*?)\r?\n---/u.exec(content);
  if (match?.[1] === undefined) {
    throw new Error("SKILL.md must start with YAML frontmatter");
  }
  const parsed = parseFlatYaml(match[1]);
  const name = parsed.name;
  const description = parsed.description;
  if (name === undefined || name.trim() === "") {
    throw new Error("frontmatter is missing name");
  }
  if (description === undefined || description.trim() === "") {
    throw new Error("frontmatter is missing description");
  }
  return { description, name };
}

function parseFlatYaml(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const line of content.split(/\r?\n/u)) {
    const trimmed = line.trim();
    if (trimmed === "" || trimmed.startsWith("#")) {
      continue;
    }
    const separator = trimmed.indexOf(":");
    if (separator === -1) {
      continue;
    }
    const key = trimmed.slice(0, separator).trim();
    const value = trimmed.slice(separator + 1).trim();
    result[key] = unquoteYamlScalar(value);
  }
  return result;
}

function unquoteYamlScalar(value: string): string {
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    return value.slice(1, -1);
  }
  return value;
}

function isDirectory(path: string): boolean {
  return existsSync(path) && statSync(path).isDirectory();
}

function defaultSkillsRoot(): string {
  return resolve(dirname(fileURLToPath(import.meta.url)), "../../");
}

function usage(message: string): never {
  console.error(message);
  console.error("Usage: tsx scripts/validate_codex_skill.ts <skill-folder-or-name>");
  process.exit(1);
}

main();
